src/api-service/__app__/onefuzzlib/proxy.py

```diff
@@ -1,271 +1,332 @@
+                return
+        self.save()
+
+    def set_provision_failed(self, vm_data: VirtualMachine) -> None:
+        errors = ["provisioning failed"]
+        for status in vm_data.instance_view.statuses:
+            if status.level.name.lower() == "error":
+                errors.append(
+                    f"code:{status.code} status:{status.display_status} "
+                    f"message:{status.message}"
+                )
+
+        self.set_failed(
+            Error(
+                code=ErrorCode.PROXY_FAILED,
+                errors=errors,
+            )
+        )
+        return
+
+    def set_failed(self, error: Error) -> None:
+        if self.error is not None:
+            return
+
+        logging.error(PROXY_LOG_PREFIX + "vm failed: %s - %s", self.region, error)
+        send_event(
+            EventProxyFailed(region=self.region, proxy_id=self.proxy_id, error=error)
+        )
+        self.error = error
+        self.state = VmState.stopping
+        self.save()
+
+    def extensions_launch(self) -> None:
+        vm = self.get_vm()
+        vm_data = vm.get()
+        if not vm_data:
+            self.set_failed(
+                Error(
+                    code=ErrorCode.PROXY_FAILED,
+                    errors=["azure not able to find vm"],
+                )
+            )
+            return
+
+        if vm_data.provisioning_state == "Failed":
+            self.set_provision_failed(vm_data)
+            return
+
+        ip = get_public_ip(vm_data.network_profile.network_interfaces[0].id)
+        if ip is None:
+            self.save()
+            return
+        self.ip = ip
+
+        extensions = proxy_manager_extensions(self.region, self.proxy_id)
+        result = vm.add_extensions(extensions)
+        if isinstance(result, Error):
+            self.set_failed(result)
+            return
+        elif result:
+            self.state = VmState.running
+
+        self.save()
+
+    def stopping(self) -> None:
+        vm = self.get_vm()
+        if not vm.is_deleted():
+            logging.info(PROXY_LOG_PREFIX + "stopping proxy: %s", self.region)
+            vm.delete()
+            self.save()
+        else:
+            self.stopped()
+
+    def stopped(self) -> None:
+        logging.info(PROXY_LOG_PREFIX + "removing proxy: %s", self.region)
+        self.delete()
+
+    def is_outdated(self) -> bool:
+        if self.version != __version__:
+            logging.info(
+                PROXY_LOG_PREFIX + "mismatch version: proxy:%s service:%s state:%s",
+                self.version,
+                __version__,
+                self.state,
+            )
+            self.outdated = True
+            return True
+        if self.created_timestamp is not None:
+            proxy_timestamp = self.created_timestamp
+            if proxy_timestamp < (
+                datetime.datetime.now(tz=datetime.timezone.utc)
+                - datetime.timedelta(days=7)

```

Comment:
Can you make this timedelta a constant at the top of the file?  (like PROXY_IMAGE).

This will help with readability in the future.

Reply:
Done.